addons/udes_stock_refactoring/models/stock_move.py

```diff
@@ -0,0 +1,334 @@
+
+        Odoos move._action_confirm returns all the moves passed in to it, after
+        merging any it can. In places the return value is used to immediately
+        assign, so any created moves should be returned.
+
+        However, in stock.move._split() in Odoo core, _action_confirm is
+        called on a single move and it expects a single move to be returned
+        (will error out if not). Therefore the refactor at confirm should avoid
+        splitting/creating moves. Luckily this is unlikely to be an issue as
+        it makes little sense to split moves at this stage, before they have
+        stock reserved against them.
+        """
+        Module = self.env["ir.module.module"]
+
+        res = super(StockMove, self)._action_confirm(*args, **kwargs)
+        post_refactor_moves = res.exists()._action_refactor(stage="confirm")
+
+        if Module.is_module_installed("mrp") and post_refactor_moves != res:
+            raise UserError(
+                _(
+                    "Post confirm refactor has created or destroyed "
+                    "moves, which could break things if you have the "
+                    "MRP module installed."
+                )
+            )
+        return res
+
+    def _action_assign(self):
+        """
+        Extend _action_assign to trigger refactor action.
+        n.b. _action_assign does not return anything in core Odoo, so we
+        don't return any extra moves that may have been created
+        by refactoring.
+        """
+        res = super(StockMove, self)._action_assign()
+
+        self.exists()._action_refactor(stage="assign")
+        return res
+
+    def _action_done(self, cancel_backorder=False):
+        """
+        Extend _action_done to trigger refactor action, and push from drop
+
+        Odoo returns completed moves.
+        Therefore we will keep track of moves created by the refactor and
+        return them as part of the set of completed moves.
+        """
+        done_moves = super(StockMove, self)._action_done()
+
+        post_refactor_done_moves = done_moves.exists()._action_refactor(stage="validate")
+        return post_refactor_done_moves
+
+    def group_by_key(self):
+        """Check each picking type has a move key format set and return the groupby"""
+
+        # TODO MTC: This can be refactored and abstracted at some point with
+        # this with the equivalent for move_line_key
+        # we need to think of how we want to do it
+
+        if any(
+            picking_type.u_move_key_format is False
+            for picking_type in self.picking_id.picking_type_id
+        ):
+            raise UserError(
+                _("Cannot group moves when their picking type has no grouping key set.")
+            )
+
+        # force recompute on u_grouping_key so we have an up-to-date key:
+        return self.with_context(compute_key=True).groupby(lambda ml: ml.u_grouping_key)
+
+    def refactor_by_move_groups(self, groups):
```

Comment:
Is this only meant to be used inside the new group by move key class? If so, shouldn't this function be in the class?

Reply:
Same for move_line.